docs/jax-101/02-jitting.ipynb

```diff
@@ -0,0 +1,570 @@
+      "\u001b[0;31mFilteredStackTrace\u001b[0m: jax.core.ConcretizationTypeError: Abstract tracer value encountered where concrete value is expected.\n\nThe problem arose with the bool function. \n\nWhile tracing the function g at <ipython-input-7-2aa78f448d5d>:3, this concrete value was not available in Python because it depends on the value of the arguments to g at <ipython-input-7-2aa78f448d5d>:3 at flattened positions [1], and the computation of these values is being staged out (that is, delayed rather than executed eagerly).\n\nYou can use transformation parameters such as static_argnums for jit to avoid tracing particular arguments of transformed functions, though at the cost of more recompiles.\n\nSee https://jax.readthedocs.io/en/latest/faq.html#abstract-tracer-value-encountered-where-concrete-value-is-expected-error for more information.\n\nEncountered tracer value: Traced<ShapedArray(bool[])>with<DynamicJaxprTrace(level=0/1)>\n"
+     ]
+    }
+   ],
+   "source": [
+    "# While loop conditioned on x and n.\n",
+    "\n",
+    "def g(x, n):\n",
+    "  i = 0\n",
+    "  while i < n:\n",
+    "    i += 1\n",
+    "  return x + i\n",
+    "\n",
+    "g_jit = jax.jit(g)\n",
+    "g_jit(10, 20)  # Should raise an error. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "isz2U_XX_wH2"
+   },
+   "source": [
+    "As you see, running the examples above will give an error. \n",
+    "\n",
+    "The problem is that we tried to condition on the *value* of an input to the function being jitted. The reason we can't do this is related to the fact mentioned above that jaxpr depends on the actual values used to trace it. \n",
+    "\n",
+    "The more specific information about the values we use in the trace, the more we can use standard Python control flow to express ourselves. However, being too specific means we can't reuse the same traced function for other values. JAX solves this by tracing at different levels of abstraction for different purposes.\n",
+    "\n",
+    "For `jax.jit`, the default level is `ShapedArray` -- that is, each tracer has a concrete shape (which we're allowed to condition on), but no concrete value. This allows the compiled function to work on all possible inputs with the same shape -- the standard use case in machine learning. However, because the tracers have no concrete value, if we attempt to condition on one, we get the error above.\n",
+    "\n",
+    "In `jax.grad`, the constraints are more relaxed, so you can do more. If you compose several transformations, however, you must satisfy the constraints of the most strict one. So, if you `jit(grad(f))`, `f` mustn't condition on value. For more detail on the interaction between Python control flow and JAX, see [this section](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#%F0%9F%94%AA-Control-Flow) of the JAX docs.\n",
```

Comment:
Suggestion:

- a11y
- "this section... of the JAX docs" -> "this section... of 🔪 JAX - The Sharp Bits 🔪 (it's quite a popular guide AFAIU! 👏 

```suggestion
    "In `jax.grad`, the constraints are more relaxed, so you can do more. If you compose several transformations, however, you must satisfy the constraints of the most strict one. So, if you `jit(grad(f))`, `f` mustn't condition on value. For more detail on the interaction between Python control flow and JAX, check [this section](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#%F0%9F%94%AA-Control-Flow) in 🔪 JAX - The Sharp Bits 🔪.\n",
```

(Emojis should be supported in notebooks 🤔)